components/ruler/index.vue

```diff
@@ -0,0 +1,357 @@
+
+      if (value <= realMin) {
+        return 0
+      } else if (value >= realMax) {
+        return unitCount * blank
+      } else {
+        return Math.ceil((value - realMin) / unit) * blank
+      }
+    },
+
+    $_draw(left) {
+      left = +left.toFixed(2)
+      const {ctx, ratio, scrollingX, canvasWidth, clientHeight} = this
+
+      this.scrollingX = left
+      this.x += scrollingX - left
+
+      // clear canvas
+      const scale = ratio * ratio
+      ctx.clearRect(0, 0, canvasWidth * scale, clientHeight * scale)
+
+      this.$_drawLine()
+    },
+
+    $_drawLine() {
+      const {ctx, x, scope, step, unit, ratio, blank, unitCount} = this
+      const {blankLeft, blankRight, canvasWidth} = this
+      const [scopeLeft] = scope
+
+      const _y = 120
+      const _fontSize = 22
+      const _stepUnit = Math.round(step / unit)
+
+      ctx.lineWidth = 2
+      ctx.font = `${_fontSize * ratio}px DINPro-Medium`
+
+      for (let i = 0; i <= unitCount; i++) {
+        const _x = x + i * blank
+
+        if (_x < 0 || _x > canvasWidth * 2) {
+          continue
+        }
+
+        // over range use another color
+        const outRange = _x < x + blankLeft || _x > x + 1 + unitCount * blank - blankRight
+        if (outRange) {
+          ctx.fillStyle = '#E2E4EA'
+          ctx.strokeStyle = '#E2E4EA'
+        } else {
+          ctx.fillStyle = '#C5CAD5'
+          ctx.strokeStyle = '#858B9C'
+        }
+
+        ctx.beginPath()
+        ctx.moveTo(_x, _y)
+
+        if (i % _stepUnit === 0) {
+          // draw text
+          const text = scopeLeft + unit * i
+          const textOffset = String(text).length * _fontSize / 2
+          ctx.fillText(text, _x - textOffset, _fontSize * ratio)
+
+          // draw line
+          ctx.lineTo(_x, _y - 40)
+        } else {
+          ctx.lineTo(_x, _y - 20)
+        }
+        ctx.stroke()
+      }
+
+      // draw base line
+      ctx.strokeStyle = '#E2E4EA'
+      ctx.beginPath()
+      ctx.moveTo(x, _y)
+      ctx.lineTo(x + unitCount * blank, _y)
+      ctx.stroke()
+
+      this.$_updateValue()
+    },
+
+    $_startDrag(event) {
+      if (this.isDragging) {
+        return
+      }
+
+      event.preventDefault()
+      event.stopPropagation()
+      this.scroller.doTouchStart(event.touches, event.timeStamp)
+
+      this.isDragging = true
+      this.isScrolling = true
+
+      window.addEventListener('mousemove', this.$_onDrag)
+      window.addEventListener('touchmove', this.$_onDrag)
+    },
+
+    $_onDrag(event) {
+      event.preventDefault()
+      event.stopPropagation()
+      if (!this.isDragging) {
+        return
+      }
+      this.scroller.doTouchMove(event.touches, event.timeStamp, event.scale)
+    },
+
+    $_stopDrag(event) {
+      event.preventDefault()
+      event.stopPropagation()
+      this.isDragging = false
+
+      this.scroller.doTouchEnd(event.timeStamp)
+
+      window.removeEventListener('mousemove', this.$_onDrag)
```

Comment:
If this component is only for mobile, there is no need `mousemove` event. If you insist, attach `mouseup` event listener as well.